File: cpp/144.Binary Tree Preorder Traversal.cpp
/**
 * Source: LeetCode  144. Binary Tree Preorder Traversal
 * Author: ceezyyy
 * Date: 10-03-2019
 */


/*
Medium

Given a binary tree, return the preorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
	\
	 2
	/
   3

Output: [1,2,3]

Follow up: Recursive solution is trivial, could you do it iteratively?

*/


#include<vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<int> preorderTraversal(TreeNode* root) {
		vector<int> res;  // store the result
		POT_Recursion(root, res);
		return res;
	}

	//preorderTraversal_Recursion
	void POT_Recursion(TreeNode *node, vector<int> &r) {
		if (!node)
			return;
		r.push_back(node->val);
		POT_Recursion(node->left, r);
		POT_Recursion(node->right, r);
	}
};


/**
 * Runtime: 4 ms, faster than 58.93% of C++ online submissions for Binary Tree Preorder Traversal.
 * Memory Usage: 9.4 MB, less than 44.83% of C++ online submissions for Binary Tree Preorder Traversal.
 */

File: Python3/144.Binary Tree Preorder Traversal.py
"""
@Title: 144. Binary Tree Preorder Traversal
@Tag: binary tree
@Date: Mar-03 2020
@Author: ceezyyy
@Difficulty: Medium
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res, stack = [], [(root, False)]
        while stack:
            node, visited = stack.pop()  # the last element
            if node:
                if visited:
                    res.append(node.val)
                else:  # preorder: root -> left -> right
                    stack.append((node.right, False))
                    stack.append((node.left, False))
                    stack.append((node, True))
        return res


"""
Runtime: 20 ms, faster than 96.33% of Python3 online submissions for Binary Tree Preorder Traversal.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Preorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
